cmd/kubeadm/app/cmd/alpha/certs.go

```diff
@@ -82,9 +87,157 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+	var config genCSRConfig
+	config.kubeadmConfig, err = configutil.LoadOrDefaultInitConfiguration(
+		o.kubeadmConfigPath,
+		&kubeadmapiv1beta2.InitConfiguration{},
+		&kubeadmapiv1beta2.ClusterConfiguration{},
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to load kubeadm config")
+	}
+	// --cert-dir takes priority over kubeadm.conf if set.
+	if o.certDir != "" {
+		config.kubeadmConfig.CertificatesDir = o.certDir
+	}
+	config.kubeConfigDir = o.kubeConfigDir
+	return &config, nil
+}
+
+// newCmdGenCSR returns cobra.Command for generating keys and CSRs
+func newCmdGenCSR() *cobra.Command {
+	flagValues := newGenCSRFlags()
+
+	cmd := &cobra.Command{
+		Use:     "generate-csr",
+		Aliases: []string{"generate-keys-and-certificate-signing-requests"},
+		Short:   "Generate keys and certificate signing requests",
+		Long: cmdutil.LongDesc(`
+			Generates keys and certificate signing requests (CSRs) for all the certificates required to run the control plane.
+			This command also generates partial kubeconfig files with private key data in the  "users > user > client-key-data" field,
+			and for each kubeconfig file an accompanying ".csr" file is created.
+
+			This command is designed for use in [Kubeadm External CA Mode](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#external-ca-mode).
+			It generates CSRs which you can then submit to your external certificate authority for signing.
+
+			The PEM encoded signed certificates should then be saved alongside the key files, using ".crt" as the file extension,
+			or in the case of kubeconfig files, the PEM encoded signed certificate should be base64 encoded
+			and added to the kubeconfig file in the "users > user > client-certificate-data" field.`),
+		Example: cmdutil.Examples(`
+			# The following command will generate keys and CSRs for all control-plane certificates and kubeconfig files:
+			kubeadm alpha certs generate-csr --kubeconfig-dir /tmp/etc-k8s --cert-dir /tmp/etc-k8s/pki
+			# ...and it produces the following directories and files:
+			tree /tmp/etc-k8s
+			/tmp/etc-k8s
+			├── admin.conf
+			├── admin.conf.csr
+			├── controller-manager.conf
+			├── controller-manager.conf.csr
+			├── pki
+			│   ├── apiserver.csr
+			│   ├── apiserver-etcd-client.csr
+			│   ├── apiserver-etcd-client.key
+			│   ├── apiserver.key
+			│   ├── apiserver-kubelet-client.csr
+			│   ├── apiserver-kubelet-client.key
+			│   ├── etcd
+			│   │   ├── healthcheck-client.csr
+			│   │   ├── healthcheck-client.key
+			│   │   ├── peer.csr
+			│   │   ├── peer.key
+			│   │   ├── server.csr
+			│   │   └── server.key
+			│   ├── front-proxy-client.csr
+			│   └── front-proxy-client.key
+			├── scheduler.conf
+			└── scheduler.conf.csr`),
+		Args: cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			config, err := flagValues.toConfig()
+			if err != nil {
+				return errors.Wrap(err, "failed to generate config")
```

Comment:
No need for a `Wrap` here. Just return the error as is.
Besides adding an additional message, `Wrap` would also create a stack trace and attach it to the resulting error. Chances are, that this has already been done by `toConfig` or some other function that's called by it.
In general, `Wrap` is placed when we call functions that are external to the kubeadm code base and are not generic `utils` types of functions.

[1]